central/networkpolicies/deployment/matcher.go

```diff
@@ -0,0 +1,132 @@
+package deployment
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	networkPolicyDS "github.com/stackrox/rox/central/networkpolicies/datastore"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/labels"
+)
+
+// IsolationDetails represents the isolation level of a deployment.
+type IsolationDetails struct {
+	PolicyIDs       []string
+	IngressIsolated bool
+	EgressIsolated  bool
+}
+
+// LabeledResource is likely network graph node that belongs to a cluster and a namespace, and should be
+// selectable by Pod Selectors.
+type LabeledResource interface {
+	GetClusterId() string
+	GetNamespace() string
+	GetPodLabels() map[string]string
+}
+
+// Matcher interface for generating Isolation Details for deployments based on Network Policies matched.
+type Matcher interface {
+	GetIsolationDetails(deploymentLabels LabeledResource) IsolationDetails
+}
+
+// ClusterNamespace is a helper struct to index network policies based on their cluster and namespace.
+type ClusterNamespace struct {
+	Cluster   string
+	Namespace string
+}
+
+type selectablePolicy struct {
+	PolicyID string
+	Selector labels.CompiledSelector
+	Types    []storage.NetworkPolicyType
+}
+
+type policyMatcherImpl struct {
+	netpolMap map[ClusterNamespace][]selectablePolicy
+}
+
+// BuildMatcher creates a matcher with pre-loaded Network Policies from a set of ClusterNamespace filter.
+func BuildMatcher(ctx context.Context, store networkPolicyDS.DataStore, namespaceFilter []ClusterNamespace) (Matcher, error) {
+	netpolMap, err := buildNetworkPolicies(ctx, store, namespaceFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	return &policyMatcherImpl{
+		netpolMap: netpolMap,
+	}, nil
+}
+
+func buildNetworkPolicies(ctx context.Context, store networkPolicyDS.DataStore, namespace []ClusterNamespace) (map[ClusterNamespace][]selectablePolicy, error) {
+	result := map[ClusterNamespace][]selectablePolicy{}
+	for _, clusterNs := range namespace {
+		policies, err := store.GetNetworkPolicies(ctx, clusterNs.Cluster, clusterNs.Namespace)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get policies for %v", clusterNs)
+		}
+
+		result[clusterNs] = []selectablePolicy{}
```

Comment:
any reason we need to initialize this? appending to a nil slice is allowed

Reply:
Done